Sources/NIOWebSocketClient/main.swift

```diff
@@ -0,0 +1,233 @@
+            }
+        case .end:
+            print("Closing channel.")
+            context.close(promise: nil)
+        }
+    }
+    
+    public func handlerRemoved(context: ChannelHandlerContext) {
+        print("HTTP handler removed.")
+    }
+    
+    public func errorCaught(context: ChannelHandlerContext, error: Error) {
+        print("error: ", error)
+        
+        // As we are not really interested getting notified on success or failure
+        // we just pass nil as promise to reduce allocations.
+        context.close(promise: nil)
+    }
+}
+
+// The web socket handler to be used once the upgrade has occurred.
+// One added, it sends a ping-pong round trip with "Hello World" data.
+// It also listens for any text frames from the server and prints them.
+
+private final class WebSocketPingPongHandler: ChannelInboundHandler {
+    typealias InboundIn = WebSocketFrame
+    typealias OutboundOut = WebSocketFrame
+    
+    let testFrameData: String = "Hello World"
+    
+    // This is being hit, channel active won't be called as it is already added.
+    public func handlerAdded(context: ChannelHandlerContext) {
+        print("WebSocket handler added.")
+        self.pingTestFrameData(context: context)
+    }
+
+    public func handlerRemoved(context: ChannelHandlerContext) {
+        print("WebSocket handler removed.")
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let frame = self.unwrapInboundIn(data)
+        
+        switch frame.opcode {
+        case .pong:
+            self.pong(context: context, frame: frame)
+        case .text:
+            var data = frame.unmaskedData
+            let text = data.readString(length: data.readableBytes) ?? ""
+            print("Websocket: Received \(text)")
+        case .connectionClose:
+            self.receivedClose(context: context, frame: frame)
+        case .binary, .continuation, .ping:
+            // We ignore these frames.
+            break
+        default:
+            // Unknown frames are errors.
+            self.closeOnError(context: context)
+        }
+    }
+    
+    public func channelReadComplete(context: ChannelHandlerContext) {
+        context.flush()
+    }
+
+    private func receivedClose(context: ChannelHandlerContext, frame: WebSocketFrame) {
+        // Handle a received close frame. We're just going to close.
+        print("Received Close instruction from server")
+        context.close(promise: nil)
+    }
+    
+    private func pingTestFrameData(context: ChannelHandlerContext) {
+        let frameData = ByteBuffer(string: self.testFrameData)
+        let frame = WebSocketFrame(fin: true, opcode: .ping, data: frameData)
+        context.write(self.wrapOutboundOut(frame), promise: nil)
+    }
+    
+    private func pong(context: ChannelHandlerContext, frame: WebSocketFrame) {
+        if let frameData = frame.data.getString(at: 0, length: self.testFrameData.count) {
```

Comment:
This may as well be `readString`, which will save you from the incorrect index usage you have here.

Reply:
(To call that just take a local mutable copy of the `WebSocketFrame`).